src/main/java/jewellery/inventory/service/ProductService.java

```diff
@@ -0,0 +1,260 @@
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class ProductService {
+
+    private final ProductRepository productRepository;
+    private final UserRepository userRepository;
+    private final ResourceRepository resourceRepository;
+    private final ResourceInUserRepository resourceInUserRepository;
+    private final ResourceInProductRepository resourceInProductRepository;
+    private final ResourceInUserService resourceInUserService;
+    private final UserMapper userMapper;
+
+    @Transactional
+    public ProductResponseDto createProduct(ProductRequestDto productRequestDto) {
+
+        User user = getUser(productRequestDto);
+        List<ResourceInProduct> resourcesInProducts = getResourceInProducts(user, productRequestDto.getResourcesContent());
+
+        Product product = getProduct(productRequestDto, user, resourcesInProducts);
+        productRepository.save(product);
+
+        product.setProductsContent(getProductsInProduct(productRequestDto.getProductsContent(), product));
+        productRepository.save(product);
+
+        resourcesInProducts.forEach(resourceInProduct -> resourceInProduct.setProduct(product));
+        resourceInProductRepository.saveAll(resourcesInProducts);
+
+        return mapToProductResponseDto(product);
+    }
+
+    public List<ProductResponseDto> getAllProducts() {
+        List<Product> products = productRepository.findAll();
+        return products.stream().map(this::mapToProductResponseDto).toList();
+    }
+
+    public ProductResponseDto getProduct(UUID id) {
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+
+        return mapToProductResponseDto(product);
+    }
+
+    public void deleteProduct(UUID id) {
+
+        Product product = productRepository.findById(id).orElseThrow(
+                () -> new ProductNotFoundException(id));
+
+        if (!product.isSold()) {
+            if (product.getContent() == null) {
+                List<ResourceInProduct> resourcesInProduct = product.getResourcesContent();
+                User owner = product.getOwner();
+                List<ResourceInUser> resourcesInUser = destroyResourceInProductToResourceInUser(owner, resourcesInProduct);
+
+                resourceInUserRepository.saveAll(resourcesInUser);
+
+                if (product.getProductsContent() != null) {
+                    product.getProductsContent().forEach(content -> {
+                        content.setContent(null);
+                        productRepository.save(content);
+                    });
+
+                    product.setProductsContent(new ArrayList<>());
+                    productRepository.save(product);
+                }
+
+                productRepository.deleteById(id);
+            } else {
+                throw new ProductContainsException(id);
+            }
+        } else {
+            throw new ProductIsSoldException(id);
+        }
+    }
```

Comment:
This method is long as well. Please, try to refactor.

Reply:
Done!